Add docs for `ConcurrentRedisLockStore` (#11342)
RasaHQ/rasa: docs/docs/migration-guide.mdx

how you can migrate from one version to another.

## Rasa 3.2 to 3.3

> ### ConcurrentRedisLockStore

Reviewer: What about migration of the DB? Do users need to specify a different DB number? What happens to previous locks?
These are questions users will probably have

Author: Addressed: https://github.com/RasaHQ/rasa/pull/11342/commits/4df9022f02a2697dfa56f03a5ccda4810369ff47

> [`ConcurrentRedisLockStore`](./lock-stores.mdx#concurrentredislockstore) is the new Redis implementation of the abstract interface `LockStore`.

Reviewer: ```suggestion
The [`ConcurrentRedisLockStore`](./lock-stores.mdx#concurrentredislockstore) is the new Redis implementation of the abstract interface `LockStore`.
```

This implementation is able to handle messages received in parallel for the same conversation id by different Rasa instances.

> In order to replace the use of `RedisLockStore` with `ConcurrentRedisLockStore`, specify the new parameter `concurrent_mode` as `true` in `endpoints.yml`:

Reviewer: ```suggestion
In order to switch from using the `RedisLockStore` to the `ConcurrentRedisLockStore`, specify the parameter (new in 3.3) `concurrent_mode` as `true` in `endpoints.yml`:
```

how you can migrate from one version to another.

## Rasa 3.2 to 3.3

### ConcurrentRedisLockStore

The [`ConcurrentRedisLockStore`](./lock-stores.mdx#concurrentredislockstore) is the new Redis implementation of the abstract interface `LockStore`.

> This implementation is able to handle messages received in parallel for the same conversation id by different Rasa instances.

Reviewer: ```suggestion
This lock store can handle receiving messages in parallel for the same conversation id by different Rasa instances.
```

In order to switch from using the `RedisLockStore` to the `ConcurrentRedisLockStore`, specify the parameter (new in 3.3) `concurrent_mode` as `true` in `endpoints.yml`:

```yaml-rasa
lock_store:
    type: "redis"
    concurrent_mode: true
    url: <url of the redis instance, e.g. localhost>
    port: <port of your redis instance, usually 6379>
    password: <password used for authentication>
    db: <number of your database within redis, e.g. 0>
    key_prefix: <alphanumeric value to prepend to lock store keys>
```

> Note that not specifying the `concurrent_mode` parameter or setting it to `false` will default to using the `RedisLockStore` implementation.

Reviewer: ```suggestion
Note that omitting the `concurrent_mode` parameter or setting it to `false` will default to using the `RedisLockStore`.
```

> No database migration is required when switching to the `ConcurrentRedisLockStore`, and you may use the same Redis database number.

Reviewer: ```suggestion
No database migration is required when switching to the `ConcurrentRedisLockStore`. You can use the same Redis instance and database number as you did previously when using the `RedisLockStore`.
```

> You may want to delete all the preexisting keys if using the same Redis database number.

Reviewer: Why may the user want to do that? Will it influence the behaviour in any way, or is it more to clear out room for storage?

Author: It's just to clear out room, because the old keys won't be used any longer, they become obsolete.

Reviewer: Makes sense, maybe make that connection explicit here?

Author: Done: https://github.com/RasaHQ/rasa/pull/11342/commits/2ad5a5a6ce35c52c8476ceceb6756ff117b86051

> The reason for this is because what used to be stored by the `RedisLockStore`(i.e., serialized [TicketLock](https://rasa.com/docs/rasa/reference/rasa/core/lock/#ticketlock-objects) instances) is no longer stored by the new `ConcurrentRedisLockStore`.

Reviewer: ```suggestion
There is no overlap in locks stored when using the `RedisLockStore` and the `ConcurrentRedisLockStore`, because the `RedisLockStore` persists serialized [TicketLock](https://rasa.com/docs/rasa/reference/rasa/core/lock/#ticketlock-objects) instances while
```

> The `ConcurrentRedisLockStore` instead stores individual `Ticket` instances, as well as the last issued ticket number.

Reviewer: ```suggestion
the `ConcurrentRedisLockStore` instead stores individual `Ticket` instances, as well as the last issued ticket number.
```

making it part of the previous sentence

> The mechanism for concurrent message handling of the same conversation id recreates the `TicketLock` from the persisted `Ticket` instances.

Reviewer: ```suggestion
The `ConcurrentRedisLockStore` recreates the `TicketLock` from the persisted `Ticket` instances, which allows it to handle concurrent messages from the same conversation id.
```

    key_prefix: <alphanumeric value to prepend to lock store keys>
```

Note that omitting the `concurrent_mode` parameter or setting it to `false` will default to using the `RedisLockStore`.

No database migration is required when switching to the `ConcurrentRedisLockStore`. You can use the same Redis instance and database number as you did previously when using the `RedisLockStore`.
You may want to delete all the preexisting keys if using the same Redis database number.
These former key-value items are no longer required by the `ConcurrentRedisLockStore` and the database can be cleared.

> There is no overlap in key-value items stored when using the `RedisLockStore` and the `ConcurrentRedisLockStore`, because the `RedisLockStore` persists serialized [TicketLock](https://rasa.com/docs/rasa/reference/rasa/core/lock/#ticketlock-objects) instances) while the `ConcurrentRedisLockStore` instead stores individual [`Ticket`](https://rasa.com/docs/rasa/reference/rasa/core/lock/#ticket-objects) instances, as well as the last issued ticket number.

Reviewer: ```suggestion
There is no overlap in key-value items stored when using the `RedisLockStore` and the `ConcurrentRedisLockStore`, because the `RedisLockStore` persists serialized [TicketLock](https://rasa.com/docs/rasa/reference/rasa/core/lock/#ticketlock-objects) instances while the `ConcurrentRedisLockStore` instead stores individual [`Ticket`](https://rasa.com/docs/rasa/reference/rasa/core/lock/#ticket-objects) instances, as well as the last issued ticket number.
```
